torch/masked/maskedtensor/binary.py

```diff
@@ -0,0 +1,201 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import torch
+
+from .core import _map_mt_args_kwargs, _masks_match, _tensors_match, _wrap_result, is_masked_tensor
+
+__all__ = []  # type: ignore[var-annotated]
+
+BINARY_NAMES = [
+    "add",
+    "atan2",
+    "arctan2",
+    "bitwise_and",
+    "bitwise_or",
+    "bitwise_xor",
+    "bitwise_left_shift",
+    "bitwise_right_shift",
+    "div",
+    "divide",
+    "floor_divide",
+    "fmod",
+    "logaddexp",
+    "logaddexp2",
+    "mul",
+    "multiply",
+    "nextafter",
+    "remainder",
+    "sub",
+    "subtract",
+    "true_divide",
+    "eq",
+    "ne",
+    "le",
+    "ge",
+    "greater",
+    "greater_equal",
+    "gt",
+    "less_equal",
+    "lt",
+    "less",
+    "maximum",
+    "minimum",
+    "fmax",
+    "fmin",
+    "not_equal",
+]
+
+INPLACE_BINARY_NAMES = [
+    n + "_"
+    for n in (
+        list(
+            set(BINARY_NAMES)
+            - {
+                "logaddexp",
+                "logaddexp2",
+                "equal",
+                "fmin",
+                "minimum",
+                "maximum",
+                "fmax",
+            }
+        )
+    )
+]
+
+
+def _get_at_least_one_mask(a, b):
+    if not is_masked_tensor(a) and not is_masked_tensor(b):
+        raise TypeError("At least one of `a` and `b` must be a MaskedTensor")
+    if not _masks_match(a, b):
+        raise ValueError("a and b must have matching masks")
+    if is_masked_tensor(a):
+        return a.get_mask()
+    return b.get_mask()
+
+
+def _binary_helper(fn, args, kwargs, inplace):
+    if len(kwargs) != 0:
+        raise ValueError("len(kwargs) must equal 0")
+    for a in args[2:]:
+        if torch.is_tensor(a):
+            raise TypeError("MaskedTensor binary ops do not support Tensor arguments aside from the lhs and rhs")
+
+    if not _masks_match(*args[:2]):
+        raise ValueError(
+            "Input masks must match. If you need support for this, please open an issue on Github."
+        )
+
+    if (
+        is_masked_tensor(args[0])
+        and is_masked_tensor(args[1])
+        and args[0].layout() != args[1].layout()
+    ):
+        raise ValueError(
+            "Inputs should match sparsity/density. If you need support this, please open an issue on Github."
+        )
+
+    data_args, data_kwargs = _map_mt_args_kwargs(
+        args, kwargs, lambda x: x.get_data()
+    )
+    mask_args, mask_kwargs = _map_mt_args_kwargs(
+        args, kwargs, lambda x: x.get_mask()
+    )
+
+    args0_layout = args[0].layout() if is_masked_tensor(args[0]) else args[0].layout
+    same_layout = (
+        (is_masked_tensor(args[1]) and args[1].layout() != args0_layout)
+        or (torch.is_tensor(args[1]) and args[1].layout != args0_layout)
+    )
```

Comment:
not sure why this is called `same_layout`? the condition it seems to be testing is basically the opposite, no?

Reply:
also you should be able to dispense with all the `is_masked_tensor()` branching here and just use `data_args` instead of `args`

Reply:
Sorry yeah it should be `==` instead of `!=`. And yeah that's fair, I suppose at the end of the day I only care about the `data_args`

Reply:
Regarding the 105-109 logic from above -- I think it'd be best to just remove that for now and allow for different layouts since that's the end goal anyway

Reply:
just curious, I'd have expected tests to catch the !=/== issue above, are we testing all combinations of layouts? I thought we were via parameterize but maybe I misread - we should be, though